### PR TITLE
Add support for attention softmax in BF16 such as for llama.

### PR DESCRIPTION
**Pull request by puneeshkhanna**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by dvarshney-habana — APPROVED** [review]

**Review by polisettyvarma — APPROVED** [review]

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -209,11 +218,13 @@ def gaudi_llama_model_forward(
     output_hidden_states: Optional[bool] = None,
     return_dict: Optional[bool] = None,
     token_idx: Optional[torch.Tensor] = None,
+    attn_softmax_bf16: Optional[bool] = None,
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:221]:
```suggestion
    attn_softmax_bf16: Optional[bool] = False,
```

File: optimum/habana/transformers/models/llama/modeling_llama.py [hunk 2]
```diff
@@ -352,6 +365,7 @@ def forward(
         output_hidden_states: Optional[bool] = None,
         return_dict: Optional[bool] = None,
         token_idx: Optional[torch.Tensor] = None,
+        attn_softmax_bf16: Optional[bool] = None,
```
Comment (regisss) [inline optimum/habana/transformers/models/llama/modeling_llama.py:368]:
```suggestion
        attn_softmax_bf16: Optional[bool] = False,
```

File: optimum/habana/transformers/generation/utils.py [hunk 3]
```diff
@@ -446,6 +446,11 @@ def generate(
             generation_config.max_length = generation_config.max_new_tokens + input_ids_length
         self._validate_generated_length(generation_config, input_ids_length, has_default_max_length)
 
+        # determine whether attention softmax needs to execute in lower precision
+        model_kwargs["attn_softmax_bf16"] = (
+            generation_config.attn_softmax_bf16 if hasattr(generation_config, "attn_softmax_bf16") else None
+        )
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:452]: I think we should define a `attn_softmax_bf16` field in: https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/configuration_utils.py
That way, we won't have to check here if it is in the generation config, and that will allow users to easily enable it in other contexts (e.g. model evaluation during training as in https://github.com/huggingface/optimum-habana/blob/4bf8a224b8838850e67ac85ca7fdbca91cc09d2d/optimum/habana/transformers/trainer_seq2seq.py#L44

**Review by regisss — COMMENTED** [review]
@puneeshkhanna Nice PR!
I left a couple of comments, mainly to add this new arg to `GaudiGenerationConfig`.

File: optimum/habana/transformers/generation/utils.py (hunk 3 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/utils.py:452]: Got it. Will push patch soon.

Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/utils.py]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 2 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

File: optimum/habana/transformers/models/llama/modeling_llama.py (hunk 1 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/models/llama/modeling_llama.py]: Done

**Comment by puneeshkhanna** [comment]
@regisss - I hope I have addressed all the nice review comments. Please check now.

File: optimum/habana/transformers/generation/configuration_utils.py [hunk 4]
```diff
@@ -21,3 +21,4 @@ def __init__(self, **kwargs):
 
         self.static_shapes = kwargs.get("static_shapes", None)
         self.ignore_eos = kwargs.get("ignore_eos", None)
+        self.attn_softmax_bf16 = kwargs.get("attn_softmax_bf16", None)
```
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:24]: Could you also add it in the docstring of the class?  You can reuse the help text you wrote in `run_generation.py`.

**Review by regisss — COMMENTED** [review]
Thanks for updating the PR @puneeshkhanna ! Let's also add `attn_softmax_bf16` in the docstring of the `GaudiGenerationConfig` class and then it will be good to merge.

File: optimum/habana/transformers/generation/configuration_utils.py (hunk 4 above)
Comment (puneeshkhanna) [reply optimum/habana/transformers/generation/configuration_utils.py:27]: So silly of me that I missed it. Done now.

**Review by regisss — APPROVED** [review]

**Comment by mandy-li** [comment]
@puneeshkhanna I am wondering why we use "--attn_softmax_bf16" arg to control if to run attn_softmax in bf16 or not? why not run attn_softmax_bf16 directly if the model is in bf16/hmp? why we let user control from cmd arg? Thanks

**Comment by puneeshkhanna** [comment]
> @puneeshkhanna I am wondering why we use "--attn_softmax_bf16" arg to control if to run attn_softmax in bf16 or not? why not run attn_softmax_bf16 directly if the model is in bf16/hmp? why we let user control from cmd arg? Thanks

@mandy-li  - Model code by default has explicit upcast to float32 (torch.nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32)) so even though model is run in bf16, softmax will always run in FP32. Usually that is the case with softmax in attention block in most of the models because of exponent operations. 
Hence we need a user control for it and if desired accuracy and perf can be achieved via softmax in BF16 only then this option can be enabled.
